File: 0036-valid-sudoku/0036-valid-sudoku.py
class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        '''
         0 1 2 3 4 5 6 7 8 9
        0
        1
        2
        3
        4
        5

        (2, 2): 0

        0, 0 -> 0
        1, 0 -> 3
        1, 2 -> 5
        2, 2 -> 8 = 2 * 3 + 2

        8 -> 2 = 8//3
        1 -> 3
        ''' 

        rows = defaultdict(set)
        cols = defaultdict(set)
        boxes = defaultdict(set)
        SIZE = 9

        for i in range(SIZE):
            for j in range(SIZE):
                char = board[i][j]
                if char == ".":
                    continue
                if char in rows[i] or char in cols[j] or char in boxes[(i // 3) * 3 + j // 3]:
                    return False
                rows[i].add(char)
                cols[j].add(char)
                boxes[(i // 3) * 3 + j // 3].add(char)

        return True

File: 0036-valid-sudoku/0036-valid-sudoku.cpp
class Solution {
public:
    bool isValidSudoku(vector<vector<char>>& board) {
        unordered_map<int, unordered_set<char>> row_set;
        unordered_map<int, unordered_set<char>> col_set;
        unordered_map<int, unordered_set<char>> box_set;
        int size = board[0].size();

        for (int i = 0; i < size; i++) {
            for (int j = 0; j < size; j++) {
                char ch = board[i][j];
                if (ch == '.')
                    continue;
                if (row_set[i].contains(ch) || 
                    col_set[j].contains(ch) || 
                    box_set[(i / 3) * 3 + (j / 3)].contains(ch)
                )
                    return false;
                row_set[i].insert(ch);
                col_set[j].insert(ch);
                box_set[(i / 3) * 3 + (j / 3)].insert(ch);
            }
        }

        return true;
    }
};

/*
0 1 2 3 4 5 6 7 8 

(0, 1) -> 1

ri = row / 3
ci = col / 3
ri * 3 + ci

*/
